### ISSUE TITLE
Change *-empty-line-before "after-comment" keyword behaviour to ignore shared-line comments

### ISSUE THREAD

**Issue by hudochenkov** [stylelint/stylelint#2212]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

I think `after-comment` keyword should worry about non-shared comments only. And look through shared comments. Currently there is no way for having declarations grouped together, but have empty line before first declaration if `prev().prev()` node has shared comment.

> Which rule, if any, is this issue related to?

`declaration-empty-line-before`
`custom-property-empty-line-before`
`rule-nested-empty-line-before`
`at-rule-empty-line-before`

> What CSS is needed to reproduce this issue?

```css
a {
    --sdfsf: 10px; /* sdfsf */

    color: #0098db; /* sfs */
    text-decoration: none;
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "declaration-empty-line-before": ["always", {
        "except": ["after-comment", "after-declaration"]
    }],
  }
}
```

> Which version of stylelint are you using?

`7.7.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Stylelint demo page and Sublime Text plugin.

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No.

> What did you expect to happen?

No warnings.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warning was flagged:

```
4:5 error Unexpected empty line before declaration (declaration-empty-line-before)
```

---

Following example with config above has no warning and shouldn't have in future:

```css
a {
    --sdfsf: 10px; 
    /* sdfsf */
    color: #0098db; /* sfs */
    text-decoration: none;
}
```

**Comment by jeddy3** [comment]
@hudochenkov Thanks for another good suggestion.

> I think `after-comment` keyword should worry about non-shared comments only. And look through shared comments.

We call those "shared-line" comments, and if your suggestion is to ignore them by default (i.e., as you put it, "look through" them when traversing the tree) then I agree with this change of behaviour. This new behaviour would be consistent with how shared-line comments are treated within [`comment-empty-line-before`](https://github.com/stylelint/stylelint/blob/master/lib/rules/comment-empty-line-before/README.md#comment-empty-line-before).

I believe, because these are `except` options, that this change could break builds. As such, I think we'll need to make this change in `8.0.0`. @hudochenkov There's a `v8` branch that you can pull request into if you'd like to contribute this change.

**Comment by hudochenkov** [comment]
> There's a v8 branch that you can pull request into if you'd like to contribute this change.

I'll do. Thank you.

**Comment by hudochenkov** [comment]
Affected options for reference:

```
at-rule-empty-line-before
ignore: ["after-comment"]

custom-property-empty-line-before
except: ["after-comment"]
ignore: ["after-comment"]

declaration-empty-line-before
except: ["after-comment"]
ignore: ["after-comment"]

rule-nested-empty-line-before
ignore: ["after-comment"]
```

**Comment by davidtheclark** [comment]
Fixed by #2262.
